### ISSUE TITLE
Storage size of 'sau' isn't known in lib/roles/https/client/client-handshake.c

### ISSUE THREAD

**Issue by FredrikFornstad** [warmcat/libwebsockets#1394]
Maybe I am a bit too early testing the libwebsockets master branch on esp32, but with the latest commit "http proxy: client: unix socket support" I get a compile error:

C:/msys32_new/home/fredrik/esp/lws-esp32-factory/components/libwebsockets/lib/roles/http/client/client-handshake.c:40:21: error: storage size of 'sau' isn't known
  struct sockaddr_un sau;
                     ^
C:/msys32_new/home/fredrik/esp/lws-esp32-factory/components/libwebsockets/lib/roles/http/client/client-handshake.c:197:20: error: 'AF_UNIX' undeclared (first use in this function)
   sau.sun_family = AF_UNIX;
                    ^
C:/msys32_new/home/fredrik/esp/lws-esp32-factory/components/libwebsockets/lib/roles/http/client/client-handshake.c:197:20: note: each undeclared identifier is reported only once for each function it appears in
C:/msys32_new/home/fredrik/esp/lws-esp32-factory/components/libwebsockets/lib/roles/http/client/client-handshake.c:40:21: warning: unused variable 'sau' [-Wunused-variable]
  struct sockaddr_un sau;
                     ^

**Comment by FredrikFornstad** [comment]
Seems that a possible solution(?) to the above problem is to add the following line in lib/plat/esp32/private.h : 

#include <sys/un.h>

However, now when I might be past that problem am facing a new one:

In file included from C:/msys32_new/home/fredrik/esp/esp-idf/components/lwip/include/lwip/lwip/inet.h:44:0,
                 from C:/msys32_new/home/fredrik/esp/esp-idf/components/lwip/include/lwip/lwip/netdb.h:38,
                 from C:/msys32_new/home/fredrik/esp/esp-idf/components/lwip/include/lwip/posix/netdb.h:33,
                 from C:/msys32_new/home/fredrik/esp/lws-esp32-factory/components/libwebsockets/lib/plat/esp32/private.h:33,
                 from C:/msys32_new/home/fredrik/esp/lws-esp32-factory/components/libwebsockets/lib/core/private.h:109,
                 from C:/msys32_new/home/fredrik/esp/lws-esp32-factory/components/libwebsockets/lib/roles/http/client/client-handshake.c:1:
C:/msys32_new/home/fredrik/esp/lws-esp32-factory/components/libwebsockets/lib/roles/http/client/client-handshake.c: In function 'lws_client_connect_2':
C:/msys32_new/home/fredrik/esp/esp-idf/components/lwip/include/lwip/lwip/def.h:86:18: error: 'port' may be used uninitialized in this function [-Werror=maybe-uninitialized]
 #define htons(x) lwip_htons(x)
                  ^
C:/msys32_new/home/fredrik/esp/lws-esp32-factory/components/libwebsockets/lib/roles/http/client/client-handshake.c:44:9: note: 'port' was declared here
  int n, port;
         ^

Not really sure what is going on here (and yes, I am trying to build with the latest esp-idf master commit)

**Comment by lws-team** [comment]
> Maybe I am a bit too early testing the libwebsockets master branch on esp32

No, it's great.

> However, now when I might be past that problem am facing a new one:

The latest patch on lws is about supporting reverse proxying by unix sockets... so a mount can actually be an http or https link over a local system socket that lws proxies the requests to transparently.

I think generally no point on ESP32 to separate server processes (it has no process concept) and have them communicate by unix sockets.  Lwip supports them evidently but it will have had no testing in esp-idf.

The problem is coming because the patch changes lws cmake to have that on by default, since in many cases, even on weak embedded systems, it does make sense.  This should solve it

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index c5a661a4..8ef78211 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -136,7 +136,7 @@ endif()
 
 # microsoft... that's why you can't have nice things
 
-if (WIN32)
+if (WIN32 OR LWS_WITH_ESP32)
        set(LWS_UNIX_SOCK 0)
 endif()
 
```

**Comment by FredrikFornstad** [comment]
Can confirm that the patch to CMakeLists.txt allows it to compile as you suggested.

**Comment by FredrikFornstad** [comment]
Sorry, not sure what I did… It does NOT compile with your patch. It still complains about sau storage size not known and that AF_UNIX is not declared.

Shouldn't there be some "#ifdef LWS_WITH_UNIX_SOCK" guards around the sau declaration and a couple of more places (somewhere around line 40, 195 and 445) in libwebsockets/lib/roles/http/client/client-handshake.c as you have it for the server.c?

**Comment by FredrikFornstad** [comment]
Great! It now builds OK (at least using a 3 day old esp-idf). 

I can see that espressif yesterday moved lwip into a sub-module and I have a feeling that the path info to the compiler might need some updates due to this. I will try it out soon and raise a new issue/pull request if I see any problems.
